File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.java
class Solution {
    boolean help(int idx, int[] nums, int target, int[][] dp){

        if(target == 0){
        return true;
    }
    if(target != 0 && idx == nums.length) return false;    
    if(idx == nums.length || target <0) return false;
    
    if(dp[idx][target] != -1) return dp[idx][target] == 1;

  boolean take =  help(idx+1,nums,target-nums[idx],dp);
  boolean ntake =  help(idx+1,nums,target,dp);

  dp[idx][target] = (take || ntake) ? 1: 0;

  return take || ntake;
    }

    public boolean canPartition(int[] nums) {
       int sum = 0;
       for(int i =0;i<nums.length;i++){
        sum += nums[i];
       }
       if(sum%2 != 0) return false;
       int target = sum/2;
       int[][] dp = new int[nums.length][target+1];
       for(int[] a: dp){
         Arrays.fill(a,-1);
       }
       return help(0,nums,target,dp);



    }
}

File: 0416-partition-equal-subset-sum/0416-partition-equal-subset-sum.cpp
class Solution {
public:

bool helper(int idx , int sum1 , int total,vector<int>& arr,vector<vector<int>>& dp){
   if(sum1 == (total - sum1)) return true;

   if(idx == arr.size()-1){
     if((sum1+arr[idx]) == total -(sum1+arr[idx]) ) return true;
     return false;
   }
     if(dp[idx][sum1] != -1) return dp[idx][sum1];
   bool pick = false;
   if(sum1+arr[idx] <= total){
    pick = helper(idx+1,sum1+arr[idx], total,arr,dp);
   }

   bool nopick = helper(idx+1,sum1,total,arr,dp);

   return dp[idx][sum1] = pick || nopick;
}
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for(int i =0;i<nums.size();i++){
             sum += nums[i];
        }
        vector<vector<int>>dp(nums.size(),vector<int>(sum+1,-1));

        return helper(0,0,sum,nums,dp);
    }
};